curp/src/client_new/unary.rs

```diff
@@ -0,0 +1,625 @@
+    /// Cluster version
+    version: AtomicU64,
+    /// Cluster connects
+    connects: DashMap<ServerId, Arc<dyn ConnectApi>>,
+}
+
+impl Debug for ClusterState {
+    #[inline]
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ClusterState")
+            .field("version", &self.version)
+            .field(
+                "connects",
+                &self.connects.iter().map(|v| *v.key()).collect_vec(),
+            )
+            .finish()
+    }
+}
+
+impl ClusterState {
+    /// Check the cluster version and update the connects if needed.
+    async fn check_and_update(
+        &self,
+        cluster_version: u64,
+        member_addrs: HashMap<ServerId, Vec<String>>,
+        leader_update: impl Fn(),
+    ) -> Result<(), tonic::transport::Error> {
+        // Check and update the cluster version in client side
+        // FIXME: is these ordering options correct?
+        let update = self.version.fetch_update(
+            std::sync::atomic::Ordering::Acquire,
+            std::sync::atomic::Ordering::Acquire,
+            |cur| (cluster_version > cur).then_some(cluster_version),
+        );
+        if update.is_err() {
+            return Ok(());
+        }
+        // TODO: if the above code can ensure that only one thread enters (aka. optimistic lock),
+        // can `DashMap` be replaced with some unsafe codes with &HashMap to improve performance?
+        let old_ids = self
+            .connects
+            .iter()
+            .map(|c| *c.key())
+            .collect::<HashSet<_>>();
+        let new_ids = member_addrs.keys().copied().collect::<HashSet<_>>();
+        let diffs = &old_ids ^ &new_ids;
+        for diff in diffs {
+            if self.connects.contains_key(&diff) {
+                debug!("client removes old server({diff})");
+                let _ig = self.connects.remove(&diff);
+            } else {
+                let addrs = member_addrs
+                    .get(&diff)
+                    .unwrap_or_else(|| unreachable!("{diff} must in new member addrs"))
+                    .clone();
+                debug!("client connects to a new server({diff}), address({addrs:?})");
+                let new_conn = rpc::connect(diff, addrs).await?;
+                let _ig = self.connects.insert(diff, new_conn);
+            }
+        }
+        let sames = &old_ids & &new_ids;
+        for same in sames {
+            let conn = self
+                .connects
+                .get(&same)
+                .unwrap_or_else(|| unreachable!("{same} must in old connects"));
+            let addrs = member_addrs
+                .get(&same)
+                .unwrap_or_else(|| unreachable!("{same} must in new member addrs"))
+                .clone();
+            conn.update_addrs(addrs).await?;
+        }
+        // Update the leader state here to ensure that each update of the cluster
+        // state and the leader state are atomic.
+        leader_update();
```

Comment:
The entire `check_and_update` does not seems to be atomic, maybe we need an `RwLock` for all the states? We rarely need to update the states so I think performance is not critcal here.

Reply:
L110~117 should form an optimistic lock, and only a specific thread with the highest version can enter the subsequent update process; the others will return Ok(()) early.

```rust
        let update = self.version.fetch_update(
            std::sync::atomic::Ordering::Acquire,
            std::sync::atomic::Ordering::Acquire,
            |cur| (cluster_version > cur).then_some(cluster_version),
        );
        if update.is_err() {
            return Ok(());
        }
```

Reply:
> If the current `cluster_version` is 1, and two threads call this method with 2 and 3 respectively, then both 1->2 and 2->3 changes may be successful, and the subsequent update process will be executed concurrently.This doesn't seem to guarantee that only one thread will go to the subsequent update process.

Yes, this issue still exists; programming without locks always tends to confuse me.

Reply:
It is refactored now